k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -0,0 +1,133 @@
+package io.buoyant.k8s
+
+import com.twitter.concurrent.AsyncStream
+import com.twitter.finagle.http
+import com.twitter.finagle.http.Response
+import com.twitter.finagle.param.HighResTimer
+import com.twitter.finagle.service.{Backoff, RetryBudget, RetryFilter, RetryPolicy}
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.io.Reader
+import com.twitter.util._
+import com.twitter.util.TimeConversions._
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * An abstract class that encapsulates the ability to Watch a k8s [[Resource]].
+ */
+private[k8s] abstract class Watchable[O <: KubeObject: Manifest, W <: Watch[O]: Manifest]
+  extends Resource {
+  import Watchable._
+
+  protected def backoffs: Stream[Duration]
+  protected def path: String
+  protected def stats: StatsReceiver
+
+  /**
+   * implementing classes should define this method to retrieve the current state
+   * of the resource to pass as Modified results to any watchers.
+   *
+   * @return a Future containing a sequence of Watches and an optional String representing the current resourceVersion
+   */
+  protected def restartWatches(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None
+  ): Future[(Seq[W], Option[String])]
+
+  protected def infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
+    RetryPolicy.backoff(backoffs) {
+      // We will assume 5xx are retryable, everything else is not for now
+      case (_, Return(rep)) => rep.status.code >= 500 && rep.status.code < 600
+      case (_, Throw(NonFatal(ex))) =>
+        log.error(s"retrying k8s endpoints request on error $ex")
```

Comment:
i think this log message is no longer correct? this may not be an endpoints request